### ISSUE TITLE
Thoughts on licensing and package layout

### ISSUE THREAD

**Issue by jtdaugherty** [jgm/skylighting#35]
In #34 I mentioned that Skylighting would be able to be licensed more permissively (e.g. BSD3) if it were to read and parse the syntax definition files rather than using them as the basis for code generation. That makes me wonder whether you would be willing to consider breaking Skylighting up into two packages as follows:

* `skylighting-core`, which provides the data types and parsing/lexing helper code and styles, support for reading XML syntax definitions and using them to parse at runtime, but no concrete parsers. This would be licensed BSD3 and would not ship any XML files or code generated from them.
* `skylighting`, which depends on `skylighting-core` and ships code generated from the XML files that it currently uses. This would be licensed GPL accordingly.

This kind of break-up would permit projects wanting to use Skylighting with a more permissive license to do so, perhaps at a performance cost due to runtime XML processing. I don't know the Skylighting code well enough to know whether there is other work that would have to be done to make this kind of split possible.

What do you think about that kind of arrangement? Is that something you'd be willing to do? If not, would you be willing to accept a patch for it?

CC @kquick

**Comment by jgm** [comment]
I don't think I have any objection to that.  There are a few
other contributors besides me who would also need to be
consulted.

I also haven't thought through whether there'd be any
unanticipated technical problems. You can go ahead and try,
though.

+++ Jonathan Daugherty [Jan 27 18 00:21 ]:

**Comment by jtdaugherty** [comment]
Okay, thanks for considering it. As for the other contributors, should I approach them? Or will you? I can start working on a patch this week.

**Comment by jtdaugherty** [comment]
Also, I'll amend my original proposal and say that `skylighting-core` *could* ship the XML files and include licensing information accordingly without itself needing to be licensed under the GPL. That way, users of just the `core` library could still get access to the XML definitions.

**Comment by jgm** [comment]
There are really only two other contributors, besides me,
who have contributed real code (I disregard trivial typo
fixes and changes to the build or CI system):

@pthariensflame
@dbaynard

Perhaps they can chime in on whether they'd object to
having the core skylighting code relicensed as BSD3.
I'd be happy with this, myself.

+++ Jonathan Daugherty [Jan 28 18 21:07 ]:
>   Okay, thanks for considering it. As for the other contributors, should
>   I approach them? Or will you? I can start working on a patch this week.

**Comment by dbaynard** [comment]
Fine by me.

**Comment by jtdaugherty** [comment]
If I work on a patch for this, how would you prefer it to look in the end? The options that come to mind for me are
* this repository contains directories `skylighting` and `skylighting-core` with the aforementioned packages, or
* this repository continues to correspond to the `skylighting` package but a new repository, `skylighting-core`, adopts sources from this tree.

**Comment by jtdaugherty** [comment]
Also, where the styles in `Styles.hs` derived from GPL'd theme files?

**Comment by pthariensflame** [comment]
For my part, I'd be *all for* a BSD relicense.

**Comment by jgm** [comment]
+++ Jonathan Daugherty [Jan 28 18 18:39 ]:
>   Also, where the styles in Styles.hs derived from GPL'd theme files?

Hm, I'm not sure.

I believe the 'kate' style is from default.theme from KDE,
but I don't know the license terms for that.

breezeDark is also from a KDE theme I found somewhere.

The others are my own attempts to match colors from other
frameworks, none are automatically derived.

**Comment by jgm** [comment]
The single repository option is probably the best, since
the history will remain continuous.

+++ Jonathan Daugherty [Jan 28 18 18:24 ]:

**Comment by jtdaugherty** [comment]
Okay. Is it safe to say that we shall assume that the themes are BSD-licensed, then? To put it another way: shall we assume that the themes can go in the `core` package?

**Comment by jgm** [comment]
+++ Jonathan Daugherty [Jan 28 18 23:01 ]:
>   Okay. Is it safe to say that we shall assume that the themes are
>   BSD-licensed, then? To put it another way: shall we assume that the
>   themes can go in the core package?

You may need to hunt them down to find out for sure.

**Comment by jtdaugherty** [comment]
Another question: what would you like the `skylighting-core` version number to be initially? It could just mimic the version of `skylighting` itself to start with, to make the correspondence easier to manage going forward.

**Comment by jtdaugherty** [comment]
(I'm assuming here that after this split, the version progression of `skylighting` should continue on as before for continuity for existing users.)

**Comment by jtdaugherty** [comment]
As I work on this change, I'm running into an issue. My original thought was to move the core of the library to a `-core` package and leave the generated syntax modules in `skylighting`. But one goal I think is worthwhile is to do as much as possible to ensure that users using `skylighting` don't have to do much work to upgrade after the split.

If I do what I originally planned, the module set for `skylighting` becomes very small: just `Skylighting.Syntax`. That starts to look awfully strange given that the name of the package didn't change, and even if the split is documented it will not be very clear how to fix your program once you start failing to import modules moved to `-core`.

So I'm actually thinking now that keeping the "core" of the library in `skylighting` and *moving the syntax modules* to a new package instead is a better plan: that way the only thing that disappears from the API is stuff from `Skylighting.Syntax`, and we can have a new (GPL'd) package called `skylighting-syntax` (say) that provides the generated syntax modules. This creates less of a discontinuity for users of the `skylighting` package if they're importing and using basically any other modules from it, including the very important `Skylighting.hs`.

What do you think?

**Comment by dbaynard** [comment]
@jtdaugherty what about [reexported modules](http://cabal.readthedocs.io/en/latest/developing-packages.html#pkg-field-library-reexported-modules) supported for Cabal >= 1.22 and ghc >= 7.10?

This leaves 7.8. But as pandoc no longer tests against 7.8, does skylight need to?

**Comment by jtdaugherty** [comment]
@dbaynard interesting, thanks! I didn't know about that feature. `skylighting.cabal` currently appears to work with GHC before 7.10, at least judging by the use of `Prelude.hs` for GHCs older than 7.10, so using this would require that to be changed. (Maybe those are just cruft; I don't know.)

But yeah - if we used that feature, that would enable users of `skylighting` to continue on as if nothing happened, which would be very nice.

**Comment by dbaynard** [comment]
I was going by the travis setup. I've little experience with reexports (from memory I tried to use them for a custom prelude but hpack didn't support them at the time).

It could be done without dropping support for 7.8, I guess, and users of 7.8 would need to import `skylighting-core`.

**Comment by jtdaugherty** [comment]
I'm a fan of dropping support for old GHCs if it means less maintenance pain, but I don't know what @jgm's policy is here or with related packages, or whether there are known users running GHCs that old.

**Comment by jtdaugherty** [comment]
(Known users of `skylighting`.)

**Comment by dbaynard** [comment]
[Libraries look fine.](http://packdeps.haskellers.com/reverse/skylighting) Applications are another matter.

In any case, I hope that option helps.

**Comment by jtdaugherty** [comment]
Heh, two of those are mine. :)

**Comment by dbaynard** [comment]
Precisely

**Comment by jgm** [comment]
+++ Jonathan Daugherty [Jan 30 18 19:35 ]:
>   Another question: what would you like the skylighting-core version
>   number to be initially? It could just mimic the version of skylighting
>   itself to start with, to make the correspondence easier to manage going
>   forward.


Makes sense to me.

**Comment by jgm** [comment]
+++ Jonathan Daugherty [Jan 30 18 20:10 ]:
>   As I work on this change, I'm running into an issue. My original
>   thought was to move the core of the library to a -core package and
>   leave the generated syntax modules in skylighting. But one goal I think
>   is worthwhile is to do as much as possible to ensure that users using
>   skylighting don't have to do much work to upgrade after the split.
>
>   If I do what I originally planned, the module set for skylighting
>   becomes very small: just Skylighting.Syntax. That starts to look
>   awfully strange given that the name of the package didn't change, and
>   even if the split is documented it will not be very clear how to fix
>   your program once you start failing to import modules moved to -core.
>
>   So I'm actually thinking now that keeping the "core" of the library in
>   skylighting and moving the syntax modules to a new package instead is a
>   better plan: that way the only thing that disappears from the API is
>   stuff from Skylighting.Syntax, and we can have a new (GPL'd) package
>   called skylighting-syntax (say) that provides the generated syntax
>   modules. This creates less of a discontinuity for users of the
>   skylighting package if they're importing and using basically any other
>   modules from it, including the very important Skylighting.hs.
>
>   What do you think?

That proposal makes sense to me.  I don't think there are
too many users of skylighting yet; as long as the changelog
explains what to do, it shouldn't pose a huge problem.

There might be issues, however, about how to handle e.g.
the functions in the Skylighting base module that have to
do with the default syntax map.

So, another option would be to have skylighting include
Skylighting.Syntax and the base module Skylighting.
skylighting-core would then include the other modules,
and not the base module.

**Comment by jgm** [comment]
+++ dbaynard [Jan 31 18 04:20 ]:
>   [1]@jtdaugherty what about [2]reexported modules supported for Cabal >=
>   1.22 and ghc >= 7.10?
>
>   This leaves 7.8. But as pandoc no longer tests against 7.8, does
>   skylight need to?

I'm okay with dropping 7.8 support.

**Comment by jgm** [comment]
+1 for having skylighting simply reexport modules from
skylighting-core. Yes, I think that's an even better
solution.

**Comment by jtdaugherty** [comment]
Okay, great - thanks, @jgm and @dbaynard!

**Comment by jtdaugherty** [comment]
Over in [my fork](https://github.com/jtdaugherty/skylighting), I now have a split package arrangement working. Some details so far:

* The `skylighting` package provides `Skylighting` and `Skylighting.Syntax`, and re-exports modules from `skylighting-core`.
* The module that was `Skylighting` before is now in `skylighting-core` and is named `Skylighting.Core` to avoid a name clash with `skylighting`'s `Skylighting` module. The latter re-exports the former to ensure that the new `skylighting` package has the identical API to the old one.
* The tests, benchmarks, and `extract` program live in the `-core` package and the `skylighting` binary lives in the `skylighting` package. The `bootstrap` Cabal flag is no longer needed.
* The `-core` package provides the `xml` directory contents along with a separate GPL license file.
* Syntax code generation is done more or less the same as before - essentially,
```
$ cd skylighting
$ /path/to/skylighting-extract ../skylighting-core/xml/*.xml
```

I have more to do, such as refactoring the README so there's a top-level README in addition to per-package READMEs, updating the `Makefile`, etc. I don't use Stack so I might like some help in vetting those changes, but otherwise I should probably be able to get this done in the next few days at which point it'll be ready for review.

**Comment by jtdaugherty** [comment]
One other thing: so far the API should be the same for the new `skylighting` package, but due to changes in how the API is assembled via re-exports, the Haddock will probably change. I guess we could avoid some of that by explicitly re-exporting specific names from the `core` package in `Skylighting.hs` to get the prior Haddock layout back.

**Comment by jgm** [comment]
Re-exporting the names in the base module might be sensible,
to make it easier for users to see how to use the package.

+++ Jonathan Daugherty [Feb 01 18 05:57 ]:

**Comment by jtdaugherty** [comment]
Okay, will do. Thanks!

**Comment by jtdaugherty** [comment]
The package split in my forked repository is now ready for review. What I have done that could use some attention:

* Adjusted most targets in the `Makefile` to do the right thing in the new regime. I didn't test the ones that use `stack`, but I did adjust them in what seemed like the right way.
* Made various additions and edits to the `skylighting` README dealing with the package split and licensing changes and symlinked it to the repo root.
* Added a README for `skylighting-core` that mostly refers readers to the `skylighting` README.

Other context:
* Because the `extract` program moved to `core`, bootstrapping is no longer needed. Just build the core package, then use the resulting extraction program to generate sources in the `skylighting` package. The corresponding cabal flag went away.
* Due to using the `reexported-modules` Cabal directive, GHC 7.8 support was removed, making 7.10 the oldest GHC officially supported.
* The `Makefile` assumed that `skylighting-extract` was in the `PATH`, which if true must have been because of either `stack` behavior or the original user's environment. That assumption wasn't true for me, but nonetheless I update that target to work as long as that assumption holds.

What I have *not yet done*:
* Update the Travis build script.
* Investigate the licensing terms of the color themes.

Let me know what you think, and if there are other things I should add to my list!

**Comment by jtdaugherty** [comment]
("bootstrapping is no longer needed" == "special build instructions prior to doing extraction are no longer needed")

**Comment by jtdaugherty** [comment]
I'll also note that the instructions on how to generate sources in the README are a bit incomplete, but then again, they were before, too. Right now the `Makefile` provides the most comprehensive account of how to deal with the XML files to generate modules in `skylighting`, so if need be, we could add a bit more detail to that procedure in the README. (Although that sort of thing tends to become stale, so the README *could* just say "use the Makefile.")

**Comment by jtdaugherty** [comment]
(Link, so you don't have to hunt for it: https://github.com/jtdaugherty/skylighting)

**Comment by jgm** [comment]
I've only had time to have a very brief look.

Shouldn't the xml files themselves be in `skylighting`?  That would clarify things, I think, not to be distributing these GPL'd files as part of the BSD3 skylighting-core package.  (I'm not sure about the legality of doing so; maybe it's okay as long as this they aren't used to generate code in `skylighting-core`, but even so, is there a good reason not to put these in the `skylighting` repository?)

The README sounds like it's saying that all of the xml files are GPL-licensed ("the KDE XML files licensed under the GPL"); that could be clarified, I think, since only some are GPL.

We can do travis after this is merged, it will be easier that way.

Re the themes, I found this:
https://github.com/KDE/syntax-highlighting/commit/05e7cd86367fa2dbd83548b13050dbe3fffe8ed0
So it seems the themes are MIT licensed.

**Comment by jtdaugherty** [comment]
> Shouldn't the xml files themselves be in skylighting? That would clarify things, I think, not to be distributing these GPL'd files as part of the BSD3 skylighting-core package. (I'm not sure about the legality of doing so; maybe it's okay as long as this they aren't used to generate code in skylighting-core, but even so, is there a good reason not to put these in the skylighting repository?)

My thought was that users of `skylighting-core` (such as myself) will still want to access to the XML files, but just not in the same was as they're currently used (i.e., runtime-loaded vs. used to generate code that causes license infection). My understanding from a conversation with Galois' lawyer is that the files can be distributed under their own license as part of that package. (That doesn't mean it isn't confusing, just okay.) My thinking is that if they're bundled in `skylighting` instead, then users of `core` still have to have a way of getting those files, and users of `skylighting` aren't likely to need them because they're using bundled versions anyway via the generated syntax modules.

> The README sounds like it's saying that all of the xml files are GPL-licensed ("the KDE XML files licensed under the GPL"); that could be clarified, I think, since only some are GPL.

Okay, thanks - I wasn't aware that it was only some that were affected. That suggests to me that instead of having a license file for those (`xml/LICENSE`), we just indicate that any GPL'd XML files are marked accordingly in comments in each file. Is that sufficient?

> So it seems the themes are MIT licensed.

Okay, thanks for looking that up! I assume that means it's safe to bundle those with the `core` package, then - are you okay with that?

**Comment by jgm** [comment]
+++ Jonathan Daugherty [Feb 05 18 20:54 ]:
>     Shouldn't the xml files themselves be in skylighting? That would
>     clarify things, I think, not to be distributing these GPL'd files as
>     part of the BSD3 skylighting-core package. (I'm not sure about the
>     legality of doing so; maybe it's okay as long as this they aren't
>     used to generate code in skylighting-core, but even so, is there a
>     good reason not to put these in the skylighting repository?)
>
>   My thought was that users of skylighting-core (such as myself) will
>   still want to access to the XML files, but just not in the same was as
>   they're currently used (i.e., runtime-loaded vs. used to generate code
>   that causes license infection). My understanding from a conversation
>   with Galois' lawyer is that the files can be distributed under their
>   own license as part of that package. (That doesn't mean it isn't
>   confusing, just okay.) My thinking is that if they're bundled in
>   skylighting instead, then users of core still have to have a way of
>   getting those files, and users of skylighting aren't likely to need
>   them because they're using bundled versions anyway via the generated
>   syntax modules.

It sounds okay, then.

>     The README sounds like it's saying that all of the xml files are
>     GPL-licensed ("the KDE XML files licensed under the GPL"); that
>     could be clarified, I think, since only some are GPL.
>
>   Okay, thanks - I wasn't aware that it was only some that were affected.
>   That suggests to me that instead of having a license file for those
>   (xml/LICENSE), we just indicate that any GPL'd XML files are marked
>   accordingly in comments in each file. Is that sufficient?

Yes.  (Or you could just say that many/most of the xml files are
GPL or LGPG licensed.)
>
>     So it seems the themes are MIT licensed.
>
>   Okay, thanks for looking that up! I assume that means it's safe to
>   bundle those with the core package, then - are you okay with that?

Yes.

**Comment by jtdaugherty** [comment]
@jgm https://github.com/jtdaugherty/skylighting/commit/5df0e3c7455424790f6f9e82147e17c02e872ee9

**Comment by jtdaugherty** [comment]
@jgm Have you had any time to take a closer look at my branch? I'm curious to know whether you have any concerns or would like me to make further changes.

**Comment by jgm** [comment]
I'll try to have a look later today.

+++ Jonathan Daugherty [Feb 08 18 18:02 ]:

**Comment by jgm** [comment]
I'll put some patches here, since I can't fork a fork of my own project.
```
diff --git a/skylighting/README.md b/skylighting/README.md
index c0e3d6c..b9d3d09 100644
--- a/skylighting/README.md
+++ b/skylighting/README.md
@@ -15,12 +15,12 @@ This project is divided up into two packages:
 
  * `skylighting-core`: this provides KDE XML parsing, data types,
    and output formatters. This includes the core functionality of the
-   Skylighting project licensed under the BSD3 license, along with the
+   skylighting project licensed under the BSD3 license, along with the
    KDE XML files, some of which are licensed under the LGPL or GPL. This
    package does not provide any built-in parsers corresponding to the
    XML descriptions, however. For that, use `skylighting`.
  * `skylighting`: this exposes the `skylighting-core` API but also
-   provides bundled Haskell parser modules corresponding to the XML
+   provides bundled Haskell parser modules derived from the XML
    descriptions in the `core` package. This package is entirely licensed
    under the GPL.
 
@@ -72,9 +72,12 @@ package.
 Using cabal:
 
     cd skylighting-core
-    cabal install
-    cd skylighting && skylighting-extract -- /path/to/skylighting-core/xml/*.xml
-    cabal install
+    cabal install -fexecutable
+    # or stack install --flag skylighting-core:executable
+    cd skylighting
+    skylighting-extract /path/to/skylighting-core/xml/*.xml
+    cabal install -fexecutable
+    # or stack install --flag skylighting:executable
```

This is needed to make building of `skylighting-extract` optional (which we definitely want -- otherwise everyone who uses it as a library will get this useless executable):
```
diff --git a/skylighting-core/skylighting-core.cabal b/skylighting-core/skylighting-core.cabal
index 418bd8c..0455cb9 100644
--- a/skylighting-core/skylighting-core.cabal
+++ b/skylighting-core/skylighting-core.cabal
@@ -202,6 +202,10 @@ executable skylighting-extract
     build-depends:     regex-pcre
   else
     build-depends:     regex-pcre-builtin
+  if flag(executable)
+    buildable:         True
+  else
+    buildable:         False
   hs-source-dirs:      bin
   main-is:             extract.hs
   default-language:    Haskell2010
```

**Comment by jgm** [comment]
The README says:

> Note that both the library and the executable can dynamically load
  syntax definitions, so you may not need to compile them in. If you
  prefer this approach, you can use the `skylighting-core` package
  directly which provides the XML files without the generated code
  produced by the bootstrap process described above.

But note!  Only `skylighting` provides the `skylighting` executable, and it links in the GPL'd modules.  How to deal with this?

**Comment by jgm** [comment]
The file TODO in the root directory can be removed.
I forgot to delete it earlier.

**Comment by jgm** [comment]
Building with stack gives warning:
```
Warning: File listed in skylighting-core/skylighting-core.cabal file does not exist: xml/LICENSE
```

**Comment by jgm** [comment]
Another problem is that the test-skylighting program in skylighting-core depends on the Skylighting library, and so can't be built.

We could move the tests to skylighting.  Or, better, we could do that AND create a version of the test suite that works with skylighting-core, with just dynamically loaded modules.

**Comment by jgm** [comment]
Ditto for the benchmark-skylighting program in skylighting-core.

**Comment by jtdaugherty** [comment]
> I'll put some patches here, since I can't fork a fork of my own
> project.

Done in https://github.com/jtdaugherty/skylighting/commit/a259032af260fee3aca81f509152d9f8bbf51f4e

-- 
  Jonathan Daugherty

**Comment by jtdaugherty** [comment]
> The file TODO in the root directory can be removed.
> I forgot to delete it earlier.

Done in https://github.com/jtdaugherty/skylighting/commit/534684805af902a3ef77fcfb2bb409281865a09d

-- 
  Jonathan Daugherty

**Comment by jtdaugherty** [comment]
> Building with stack gives warning:
> ```
> Warning: File listed in skylighting-core/skylighting-core.cabal file does not exist: xml/LICENSE
> ```

Done in https://github.com/jtdaugherty/skylighting/commit/970c64595d9d95aa3c70d906ea41b95303be02ba

-- 
  Jonathan Daugherty

**Comment by jtdaugherty** [comment]
> We could move the tests to skylighting. Or, better, we could
> do that AND create a version of the test suite that works with
> skylighting-core, with just dynamically loaded modules.

I was hoping to maintain a separation that allowed `skylighting` to be
more or less purely cosmetic, with all the tests and core functionality
staying behind in `core`. We could leave the tests there, but modify
them to load the XML files instead of using the defaultSyntaxMap.

-- 
  Jonathan Daugherty

**Comment by jtdaugherty** [comment]
> The README says:
> 
> > Note that both the library and the executable can dynamically load
>   syntax definitions, so you may not need to compile them in. If you
>   prefer this approach, you can use the `skylighting-core` package
>   directly which provides the XML files without the generated code
>   produced by the bootstrap process described above.
> 
> But note! Only `skylighting` provides the `skylighting` executable,
> and it links in the GPL'd modules. How to deal with this?

I'm not sure what to say about this. That text from the README seems
like a mixture of old text with new and I didn't edit it fully. It
should probably read:

> Note that both library can dynamically load syntax definitions, so
> if you prefer not to compile them into your program for licensing
> reasons, you can use the `skylighting-core` package which provides the
> XML files without the generated code produced by the bootstrap process
> described above. On the other hand, if the GPL licensing of
> `skylighting` is compatible with your program, you can use the
> generated modules in that library.

Does that help? Basically I just wanted to avoid mentioning the
executable. I think my intention with the previous edit was just to
provide some guidance on how to choose when to use which package.

-- 
  Jonathan Daugherty

**Comment by jtdaugherty** [comment]
(Edited above.)

**Comment by dbaynard** [comment]
I now have some time to help, again. What needs doing, @jtdaugherty?

**Comment by jtdaugherty** [comment]
@dbaynard I believe the ball is in @jgm's court; I'm awaiting further feedback on my branch based on the conversation above. I've finished up the bulk of the work, but as far as I know, the last task (aside from any future feedback from others) is to update the Travis build script to work with the new package layout.

**Comment by jtdaugherty** [comment]
Hi, @jgm - do you have any more questions or feedback, or things you'd like me to take care of? Do you have a sense for when you might be able to take a deeper look at this or potentially merge it?

**Comment by jgm** [comment]
Sorry, lost track of this.  I'll merge this, and we can make further adjustments, if needed, afterwards.

**Comment by jgm** [comment]
Oh, can you make a PR?

**Comment by jtdaugherty** [comment]
Done: https://github.com/jgm/skylighting/pull/37

**Comment by jgm** [comment]
Travis fixed, tests now pass.
Still probably could use a final look-over before release.

**Comment by jtdaugherty** [comment]
Great! If there is anything I can do, let me know.

**Comment by jgm** [comment]
OK, released both packages as 0.7.

Hope everything is right, but if not, we can fix later.
I did confirm that pandoc can compile against 0.7.

+++ Jonathan Daugherty [Mar 04 18 02:10 ]:

**Comment by jtdaugherty** [comment]
Thanks so much, everyone!
